### PR TITLE
upgrade dlpack to 0.5

### PR DESCRIPTION
**Pull request by cwharris**
Update dlpack from `0.3` to `>=0.5,<0.6.0a0`, and fix for breaking changes introduced in `0.4`.

closes https://github.com/rapidsai/cudf/issues/7679

### CODE REVIEW

File: cpp/cmake/thirdparty/CUDF_GetDLPack.cmake
Comment (kkraus14) [inline cpp/cmake/thirdparty/CUDF_GetDLPack.cmake:39]: I could have sworn there's some header changes in 0.5 that require us updating a little bit of code.

We also need to update the conda recipes / environment files based on this change. The integration repo as well: https://github.com/rapidsai/integration/blob/branch-21.06/conda/recipes/versions.yaml#L55-L56

We also need to get dlpack conda packages updated in conda-forge. Working on the conda-forge updates...

Comment (cwharris) [reply cpp/cmake/thirdparty/CUDF_GetDLPack.cmake:39]: I figured there were changes necessary, but I tried a clean build and everything seemed fine. :shrug:

I'll take a closer look.

**Comment by cwharris** [comment]
waiting for conda-forge to get updated with dlpack 0.5 so I can test.

File: cpp/cmake/thirdparty/CUDF_GetDLPack.cmake
Comment (galipremsagar) [reply cpp/cmake/thirdparty/CUDF_GetDLPack.cmake:39]: Seems like we might have to update our code as `dlpack`-`0.5` seems to be breaking libcudf build:

```
[ 31%] Building CXX object CMakeFiles/cudf.dir/src/io/comp/brotli_dict.cpp.o
[ 32%] Building CXX object CMakeFiles/cudf.dir/src/io/comp/cpu_unbz2.cpp.o
[ 32%] Building CUDA object CMakeFiles/cudf.dir/src/io/comp/debrotli.cu.o
[ 32%] Building CUDA object CMakeFiles/cudf.dir/src/io/comp/gpuinflate.cu.o
[ 33%] Building CUDA object CMakeFiles/cudf.dir/src/io/comp/snap.cu.o
[ 33%] Building CXX object CMakeFiles/cudf.dir/src/io/comp/uncomp.cpp.o
[ 33%] Building CUDA object CMakeFiles/cudf.dir/src/io/comp/unsnap.cu.o
In file included from /nvme/0/pgali/cudf/cpp/include/cudf/utilities/type_dispatcher.hpp:22,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/utilities/traits.hpp:20,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/column/column_view.hpp:19,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/column/column.hpp:18,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/column/column_factories.hpp:18,
                 from /nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:16:
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp: In function 'std::unique_ptr<cudf::table> cudf::detail::from_dlpack(const DLManagedTensor*, rmm::cuda_stream_view, rmm::mr::device_memory_resource*)':
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:140:33: error: 'const struct DLTensor' has no member named 'ctx'
  140 |   CUDF_EXPECTS(kDLGPU == tensor.ctx.device_type || kDLCPU == tensor.ctx.device_type ||
      |                                 ^~~
/nvme/0/pgali/cudf/cpp/include/cudf/utilities/error.hpp:63:7: note: in definition of macro 'CUDF_EXPECTS'
   63 |   (!!(cond)) ? static_cast<void>(0)                                 \
      |       ^~~~
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:140:69: error: 'const struct DLTensor' has no member named 'ctx'
  140 |   CUDF_EXPECTS(kDLGPU == tensor.ctx.device_type || kDLCPU == tensor.ctx.device_type ||
      |                                                                     ^~~
/nvme/0/pgali/cudf/cpp/include/cudf/utilities/error.hpp:63:7: note: in definition of macro 'CUDF_EXPECTS'
   63 |   (!!(cond)) ? static_cast<void>(0)                                 \
      |       ^~~~
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:141:41: error: 'const struct DLTensor' has no member named 'ctx'
  141 |                  kDLCPUPinned == tensor.ctx.device_type,
      |                                         ^~~
/nvme/0/pgali/cudf/cpp/include/cudf/utilities/error.hpp:63:7: note: in definition of macro 'CUDF_EXPECTS'
   63 |   (!!(cond)) ? static_cast<void>(0)                                 \
      |       ^~~~
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:145:14: error: 'const struct DLTensor' has no member named 'ctx'
  145 |   if (tensor.ctx.device_type != kDLCPU) {
      |              ^~~
In file included from /nvme/0/pgali/cudf/cpp/include/cudf/utilities/type_dispatcher.hpp:22,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/utilities/traits.hpp:20,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/column/column_view.hpp:19,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/column/column.hpp:18,
                 from /nvme/0/pgali/cudf/cpp/include/cudf/column/column_factories.hpp:18,
                 from /nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:16:
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:148:25: error: 'const struct DLTensor' has no member named 'ctx'
  148 |     CUDF_EXPECTS(tensor.ctx.device_id == device_id, "DLTensor device ID must be current device");
      |                         ^~~
/nvme/0/pgali/cudf/cpp/include/cudf/utilities/error.hpp:63:7: note: in definition of macro 'CUDF_EXPECTS'
   63 |   (!!(cond)) ? static_cast<void>(0)                                 \
      |       ^~~~
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp: In function 'DLManagedTensor* cudf::detail::to_dlpack(const cudf::table_view&, rmm::cuda_stream_view, rmm::mr::device_memory_resource*)':
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:237:34: error: 'struct DLTensor' has no member named 'ctx'
  237 |   CUDA_TRY(cudaGetDevice(&tensor.ctx.device_id));
      |                                  ^~~
/nvme/0/pgali/cudf/cpp/include/cudf/utilities/error.hpp:104:33: note: in definition of macro 'CUDA_TRY'
  104 |     cudaError_t const status = (call);                            \
      |                                 ^~~~
/nvme/0/pgali/cudf/cpp/src/interop/dlpack.cpp:238:10: error: 'struct DLTensor' has no member named 'ctx'
  238 |   tensor.ctx.device_type = kDLGPU;
      |          ^~~
[ 34%] Building CUDA object CMakeFiles/cudf.dir/src/io/csv/csv_gpu.cu.o
```

**Review by galipremsagar — APPROVED** [review]
This resolves the errors I was seeing. Thanks @cwharris !

File: cpp/tests/CMakeLists.txt
Comment (mythrocks) [inline cpp/tests/CMakeLists.txt:333]: Sorry, is this change unrelated to the `dlpack` change?

**Review by mythrocks — APPROVED** [review]
:+1: LGTM!

@cwharris, could we consider updating this PR's description with a very brief description of #7679, indicating that `dmlc/dlpack` had a breaking change in 0.3-0.4?
I suspect the `git log` commit message will be more informative then.

File: cpp/tests/CMakeLists.txt
Comment (robertmaynard) [reply cpp/tests/CMakeLists.txt:333]: It was a CMake anti-pattern that sneaked into the code-base. https://github.com/rapidsai/cudf/pull/8270 also fixes it.

**Review by trxcllnt — APPROVED** [review]

**Comment by cwharris** [comment]
rerun tests

File: ci/gpu/build.sh
Comment (trxcllnt) [inline ci/gpu/build.sh:87]: Looks like the [failing jobs](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-v0.20/job/cudf/job/prb/job/cudf-cpu-cuda-build/CUDA=11.2/1033/consoleText) are using `ci/cpu/build.sh`, so I suspect we also need these two lines [here](https://github.com/rapidsai/cudf/blob/branch-21.06/ci/cpu/build.sh#L54).

**Comment by cwharris** [comment]
blocked by https://github.com/rapidsai/integration/pull/274

**Comment by kkraus14** [comment]
> blocked by [rapidsai/integration#274](https://github.com/rapidsai/integration/pull/274)

I haven't seen a passing CI with overriding the package in the CI script which is typically the check we use before merging the integration PR.

**Comment by cwharris** [comment]
@kkraus14 dlpack_test.cpp wasn't getting the updated dlpack version with the build.sh change. I'll try again, but is there a reason that teat would be treated differently?

**Comment by kkraus14** [comment]
> @kkraus14 dlpack_test.cpp wasn't getting the updated dlpack headers with the build.sh change. I'll try again, but is there a reason that test would be treated differently?

I don't believe so?

**Comment by kkraus14** [comment]
I think `device` was renamed to `DLDeviceType`: https://github.com/dmlc/dlpack/blob/main/include/dlpack/dlpack.h#L38-L63

**Review by raydouglass — APPROVED** [review]

File: cpp/libcudf_kafka/cmake/thirdparty/CUDF_KAFKA_GetCUDF.cmake
Comment (kkraus14) [inline cpp/libcudf_kafka/cmake/thirdparty/CUDF_KAFKA_GetCUDF.cmake:31]: How does someone reliably rely on a libcudf install? Should we remove the patch number from the libcudf cmake instead?

Comment (cwharris) [reply cpp/libcudf_kafka/cmake/thirdparty/CUDF_KAFKA_GetCUDF.cmake:31]: @robertmaynard is working on answering that question in 21.08, but might have an answer for you now.

Comment (trxcllnt) [reply cpp/libcudf_kafka/cmake/thirdparty/CUDF_KAFKA_GetCUDF.cmake:31]: This is to work around a [CMake bug](https://gitlab.kitware.com/cmake/cmake/-/issues/22207).

Comment (kkraus14) [reply cpp/libcudf_kafka/cmake/thirdparty/CUDF_KAFKA_GetCUDF.cmake:31]: Can we use `${CUDF_KAFKA_VERSION_PATCH}` then instead?

Comment (cwharris) [reply cpp/libcudf_kafka/cmake/thirdparty/CUDF_KAFKA_GetCUDF.cmake:31]: it's a minimum version, so is that necessary?

Comment (kkraus14) [reply cpp/libcudf_kafka/cmake/thirdparty/CUDF_KAFKA_GetCUDF.cmake:31]: Didn't realize it's a minimum, perfect then!

**Comment by jakirkham** [comment]
Guessing this will eventually run into the Python test failures we are fixing in PR ( https://github.com/rapidsai/cudf/pull/8342 ). So this may need an admin merge (though probably worth waiting to make sure those are the only failures on CI)

**Review by kkraus14 — APPROVED** [review]

**Comment by cwharris** [comment]
Looks like we're failing due to the errors which will be fixed by #8342 .

```
17:29:55 Traceback (most recent call last):
17:29:55   File "<string>", line 1, in <module>
17:29:55   File "/workspace/.conda-bld/dask-cudf_1621981592845/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/python3.8/site-packages/dask_cudf/__init__.py", line 6, in <module>
17:29:55     from . import backends
17:29:55   File "/workspace/.conda-bld/dask-cudf_1621981592845/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/python3.8/site-packages/dask_cudf/backends.py", line 9, in <module>
17:29:55     from dask.dataframe.core import get_parallel_type, make_meta, meta_nonempty
17:29:55 ImportError: cannot import name 'make_meta' from 'dask.dataframe.core' (/workspace/.conda-bld/dask-cudf_1621981592845/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/python3.8/site-packages/dask/dataframe/core.py)
```

**Comment by cwharris** [comment]
@kkraus14 Ok to do an admin merge here?

**Pull request merged**
